### PR TITLE
Customizable logs feature

### PR DESCRIPTION
**Pull request by CharityKathure**
**Description:**

Merge Log Monitor Custom Log Formatting feature into main branch
**NB:** Users can specify custom log formats for all log sources _(Event Logs, ETW, File Logs and Process Monitor Logs)_

**Sample Config to use for testing::**
```
{
  "LogConfig": {
    "logFormat": "custom",
    "sources": [
      {
        "type": "EventLog",
        "startAtOldestRecord": true,
        "eventFormatMultiLine": false,
        "channels": [
          {
            "name": "system",
            "level": "Information"
          },
          {
            "name": "application",
            "level": "Error"
          }
        ],
        "customLogFormat": "{'TimeStamp':'%TimeStamp%', 'source':'%Source%', 'Severity':'%Severity%', 'EventID':'%EventID%', 'Message':'%Message%'}|JSON"
      },
      {
        "type": "ETW",
        "eventFormatMultiLine": false,
        "providers": [
          {
            "providerName": "Microsoft-Windows-WLAN-Drive",
            "providerGuid": "DAA6A96B-F3E7-4D4D-A0D6-31A350E6A445",
            "level": "Information"
          }
        ],
        "customLogFormat": "{'TimeStamp':'%TimeStamp%', 'source':'%Source%', 'Severity':'%Severity%', 'ProviderId':'%ProviderId%', 'ProviderName':'%ProviderName%', 'EventId':'%EventId%', 'EventData':'%EventData%'}|json"
      },
      {
        "type": "File",
        "directory": "c:\\inetpub\\logs",
        "filter": "*.log",
        "includeSubdirectories": true,
        "customLogFormat": "{'message':%Message%, 'source':%Source%, 'fileName':%FileName%}"
      },
      {
        "type": "Process",
        "customLogFormat": "{'TimeStamp':'%TimeStamp%', 'source':'%Source%', 'Logline':'%Logline%'}|JSON"
      }
    ]
  }
}
```

**Sample output**
```
{"TimeStamp":"2024-07-18T21:26:22.000Z", "source":"EventLog", "Severity":"Information", "EventID":"7036", "Message":"The WinRM service entered the running state."}
{"TimeStamp":"2024-07-18T21:26:22.000Z", "source":"EventLog", "Severity":"Information", "EventID":"7036", "Message":"The State Repository Service service entered the running state."}
{"TimeStamp":"2024-07-18T21:26:22.000Z", "source":"EventLog", "Severity":"Information", "EventID":"7036", "Message":"The Time Broker service entered the running state."}
{"TimeStamp":"2024-07-18T21:26:22.000Z", "source":"EventLog", "Severity":"Information", "EventID":"7036", "Message":"The Schedule service entered the running state."}
{"TimeStamp":"2024-07-18T21:26:22.000Z", "source":"EventLog", "Severity":"Information", "EventID":"7036", "Message":"The DiagTrack service entered the running state."}
{"TimeStamp":"2024-07-18T21:26:23.000Z", "source":"Process", "Logline":"Reply from ::1: time<1ms "}
{"TimeStamp":"2024-07-18T21:26:24.000Z", "source":"Process", "Logline":"Reply from ::1: time<1ms "}
{"TimeStamp":"2024-07-18T21:26:25.000Z", "source":"Process", "Logline":"Reply from ::1: time<1ms "}
```

**Unit tests:**

image

### CODE REVIEW

File: LogMonitor/src/LogMonitor/ProcessMonitor.cpp [hunk 1]
```diff
@@ -374,3 +422,16 @@ DWORD ReadFromPipe(LPVOID Param)
...
 }
+
+std::wstring ProcessMonitor::ProcessFieldsMapping(_In_ std::wstring fileFields, _In_ void* pLogEntryData)
+{
+    std::wostringstream oss;
+    ProcessLogEntry* pLogEntry = (ProcessLogEntry*)pLogEntryData;
+
+    if (Utility::CompareWStrings(fileFields, L"TimeStamp")) oss << pLogEntry->currentTime;
+    if (Utility::CompareWStrings(fileFields, L"Source")) oss << pLogEntry->source;
+    if (Utility::CompareWStrings(fileFields, L"logLine")
```
Comment (profnandaa) [inline LogMonitor/src/LogMonitor/ProcessMonitor.cpp:433]: nit: as for the name casing, you wanted to go with PascalCase or camelCase?

File: LogMonitor/src/LogMonitor/Utility.cpp [hunk 2]
```diff
@@ -343,3 +347,94 @@ int Utility::GetWaitInterval(_In_ std::double_t waitInSeconds, _In_ int elapsedT
...
+
+        if (sub[0] != '%' && sub[sub_length - 1] != '%') {
+            j++, i++;
+        } else if (sub[0] == '%' && sub[sub_length - 1] == '%' && sub_length != 1) {
+            //valid field name found in custom log format
+            wstring fieldValue;
+            if (sourceType == L"ETW") {
+                fieldValue = EtwMonitor::EtwFieldsMapping(sub.substr(1, sub_length - 2), pLogEntry);
+            }
+            if (sourceType == L"EventLog") {
```
Comment (profnandaa) [inline LogMonitor/src/LogMonitor/Utility.cpp:392]: nit: `else if` is more idiomatic.

File: LogMonitor/src/LogMonitor/ProcessMonitor.cpp (hunk 1 above)
Comment (CharityKathure) [reply LogMonitor/src/LogMonitor/ProcessMonitor.cpp]: @profnandaa I have updated the function names to PascalCase and variable names to camelCase for consistency with the existing code style.

File: LogMonitor/src/LogMonitor/Utility.cpp (hunk 2 above)
Comment (CharityKathure) [reply LogMonitor/src/LogMonitor/Utility.cpp]: Done.

**Review by iankingori — APPROVED** [review]
LGTM

**Review by profnandaa — APPROVED** [review]

**Review by bobsira — APPROVED** [review]

**Review by TinaMor — APPROVED** [review]
There are some lint warnings flagged by GitHub. Otherwise, LGTM.

**Pull request merged**